### PR TITLE
add build recipe for the tracy profiler

### PR DESCRIPTION
**Pull request by jakebolewski**
More info here: https://github.com/wolfpld/tracy

This builds the profiler gui, instrumentation library and utilities.  The only issue on OSX is with the instrumentation library with libbacktrace, I don't think the MACHO defines are correct with cross compilation and needs someone better versed to debug but everything on linux x64 should work to get started.

### CODE REVIEW

File: T/TracyProfiler/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    BuildDependency("Xorg_compositeproto_jll"),
+    BuildDependency("Xorg_damageproto_jll"),
+    BuildDependency("Xorg_fixesproto_jll"),
+    BuildDependency("Xorg_inputproto_jll"),
+    BuildDependency("Xorg_kbproto_jll"),
+    BuildDependency("Xorg_randrproto_jll"),
+    BuildDependency("Xorg_renderproto_jll"),
+    BuildDependency("Xorg_xextproto_jll"),
+    BuildDependency("Xorg_xineramaproto_jll"),
+    BuildDependency("Xorg_xproto_jll"),
```
Comment (giordano) [inline T/TracyProfiler/build_tarballs.jl:82]: `Xorg_xorgproto_jll` should do it for all of them
```suggestion
    BuildDependency("Xorg_xorgproto_jll"),
```

**Review by giordano — COMMENTED** [review]
I briefly looked into macOS build.  I'm not sure why you selected GCC as compiler, with LLVM I can get just slightly ahead.  I get stuck at
```console
sandbox:${WORKSPACE}/srcdir/tracy-0.7.8 # make -C profiler/build/unix release 
make: Entering directory '/workspace/srcdir/tracy-0.7.8/profiler/build/unix'
make[1]: Entering directory '/workspace/srcdir/tracy-0.7.8/profiler/build/unix'
c++ -c -I/workspace/destdir/include -I/workspace/destdir/include/freetype2 -I/workspace/destdir/include/capstone  -I../../../imgui -I../../libs/gl3w -O3 -flto -std=c++17 -D"DISPLAY_SERVER_X11" -DNDEBUG -D__STDC_FORMAT_MACROS -DIMGUI_IMPL_OPENGL_LOADER_GL3W -DIMGUI_ENABLE_FREETYPE ../../../server/TracyFilesystem.cpp -o obj/release/o/o/o/../../../server/TracyFilesystem.o
In file included from ../../../server/TracyFilesystem.cpp:2:
In file included from ../../../server/TracyView.hpp:23:
../../../server/TracyWorker.hpp:365:14: error: 'shared_mutex' is unavailable: introduced in macOS 10.12
        std::shared_mutex lock;
             ^
/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root/usr/include/c++/v1/shared_mutex:180:58: note: 'shared_mutex' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS _LIBCPP_AVAILABILITY_SHARED_MUTEX shared_mutex
                                                         ^
In file included from ../../../server/TracyFilesystem.cpp:2:
In file included from ../../../server/TracyView.hpp:23:
../../../server/TracyWorker.hpp:567:10: error: 'shared_mutex' is unavailable: introduced in macOS 10.12
    std::shared_mutex& GetMbpsDataLock() { return m_mbpsData.lock; }
         ^
/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root/usr/include/c++/v1/shared_mutex:180:58: note: 'shared_mutex' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS _LIBCPP_AVAILABILITY_SHARED_MUTEX shared_mutex
                                                         ^
2 errors generated.
make[1]: *** [../../../common/unix.mk:27: obj/release/o/o/o/../../../server/TracyFilesystem.o] Error 1
make[1]: Leaving directory '/workspace/srcdir/tracy-0.7.8/profiler/build/unix'
make: *** [Makefile:7: release] Error 2
make: Leaving directory '/workspace/srcdir/tracy-0.7.8/profiler/build/unix'
```
In some other places macOS 10.14 is required.  I suspect you'll have to do something like
https://github.com/JuliaPackaging/Yggdrasil/blob/3b556643f0220f8e918938b5a203cba96beb4dc1/C/Coin-OR/SHOT/build_tarballs.jl#L18-L32

File: T/TracyProfiler/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+cd ${WORKSPACE}/srcdir/tracy*
+
+# Apply patches to disable forcing -march
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/unix_library_release.patch 
+atomic_patch -p1 ${WORKSPACE}/srcdir/patches/unix_common_make.patch
+
+if [[ "${target}" == *-apple-* ]]; then
+    CC=gcc
+    CXX=g++
+fi
```
[the remark this answers is not in the document]
Comment (jakebolewski) [reply T/TracyProfiler/build_tarballs.jl:26]: I could build the GUI and all the utilities with the updated GCC compiler, but upgrading the SDK as you linked above is a better way to go.

**Comment by giordano** [comment]
```
[19:49:14] c++ -O3 -flto -mmacosx-version-min=10.15 -std=c++17 -D"DISPLAY_SERVER_X11" -s -DNDEBUG -D__STDC_FORMAT_MACROS -DIMGUI_IMPL_OPENGL_LOADER_GL3W -DIMGUI_ENABLE_FREETYPE obj/release/o/o/o/../../../common/TracySocket.o obj/release/o/o/o/../../../common/TracySystem.o obj/release/o/o/o/../../../common/tracy_lz4.o obj/release/o/o/o/../../../common/tracy_lz4hc.o obj/release/o/o/o/../../../imgui/imgui.o obj/release/o/o/o/../../../imgui/imgui_demo.o obj/release/o/o/o/../../../imgui/imgui_draw.o obj/release/o/o/o/../../../imgui/imgui_tables.o obj/release/o/o/o/../../../imgui/imgui_widgets.o obj/release/o/o/o/../../../imgui/misc/freetype/imgui_freetype.o obj/release/o/o/o/../../../server/TracyBadVersion.o obj/release/o/o/o/../../../server/TracyColor.o obj/release/o/o/o/../../../server/TracyFilesystem.o obj/release/o/o/o/../../../server/TracyMemory.o obj/release/o/o/o/../../../server/TracyMicroArchitecture.o obj/release/o/o/o/../../../server/TracyMmap.o obj/release/o/o/o/../../../server/TracyMouse.o obj/release/o/o/o/../../../server/TracyPrint.o obj/release/o/o/o/../../../server/TracySourceContents.o obj/release/o/o/o/../../../server/TracySourceTokenizer.o obj/release/o/o/o/../../../server/TracySourceView.o obj/release/o/o/o/../../../server/TracyStackFrames.o obj/release/o/o/o/../../../server/TracyStorage.o obj/release/o/o/o/../../../server/TracyTaskDispatch.o obj/release/o/o/o/../../../server/TracyTexture.o obj/release/o/o/o/../../../server/TracyTextureCompression.o obj/release/o/o/o/../../../server/TracyThreadCompress.o obj/release/o/o/o/../../../server/TracyUserData.o obj/release/o/o/o/../../../server/TracyView.o obj/release/o/o/o/../../../server/TracyWorker.o obj/release/o/o/o/../../src/HttpRequest.o obj/release/o/o/o/../../src/imgui_impl_glfw.o obj/release/o/o/o/../../src/imgui_impl_opengl3.o obj/release/o/o/o/../../src/main.o obj/release/o/o/o/../../src/NativeWindow.o obj/release/o/o/o/../../src/ResolvService.o obj/release/o/o/o/../../src/winmain.o obj/release/o/o/o/../../src/winmainArchDiscovery.o obj/release/o/o/o/../../../nfd/nfd_common.o obj/release/o/o/o/../../../zstd/common/debug.o obj/release/o/o/o/../../../zstd/common/entropy_common.o obj/release/o/o/o/../../../zstd/common/error_private.o obj/release/o/o/o/../../../zstd/common/fse_decompress.o obj/release/o/o/o/../../../zstd/common/pool.o obj/release/o/o/o/../../../zstd/common/threading.o obj/release/o/o/o/../../../zstd/common/xxhash.o obj/release/o/o/o/../../../zstd/common/zstd_common.o obj/release/o/o/o/../../../zstd/compress/fse_compress.o obj/release/o/o/o/../../../zstd/compress/hist.o obj/release/o/o/o/../../../zstd/compress/huf_compress.o obj/release/o/o/o/../../../zstd/compress/zstdmt_compress.o obj/release/o/o/o/../../../zstd/compress/zstd_compress.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_literals.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_sequences.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_superblock.o obj/release/o/o/o/../../../zstd/compress/zstd_double_fast.o obj/release/o/o/o/../../../zstd/compress/zstd_fast.o obj/release/o/o/o/../../../zstd/compress/zstd_lazy.o obj/release/o/o/o/../../../zstd/compress/zstd_ldm.o obj/release/o/o/o/../../../zstd/compress/zstd_opt.o obj/release/o/o/o/../../../zstd/decompress/huf_decompress.o obj/release/o/o/o/../../../zstd/decompress/zstd_ddict.o obj/release/o/o/o/../../../zstd/decompress/zstd_decompress.o obj/release/o/o/o/../../../zstd/decompress/zstd_decompress_block.o obj/release/o/o/o/../../../zstd/dictBuilder/cover.o obj/release/o/o/o/../../../zstd/dictBuilder/divsufsort.o obj/release/o/o/o/../../../zstd/dictBuilder/fastcover.o obj/release/o/o/o/../../../zstd/dictBuilder/zdict.o obj/release/o/o/o/../../libs/gl3w/GL/gl3w.o obj/release/o/o/o/../../../nfd/nfd_cocoa.o -L/workspace/destdir/lib -lglfw -lfreetype -lcapstone  -lpthread -ldl -framework CoreFoundation -framework AppKit -o Tracy-release
[19:49:14] ld: warning: option -s is obsolete and being ignored
[19:49:15] ld: could not parse object file obj/release/o/o/o/../../../common/TracySocket.o: 'Unknown attribute kind (62) (Producer: 'LLVM12.0.0' Reader: 'LLVM 8.0.0svn')', using libLTO version 'LLVM version 8.0.0svn' file 'obj/release/o/o/o/../../../common/TracySocket.o' for architecture x86_64
[19:49:15] clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
[19:49:15] make[1]: *** [../../../common/unix.mk:61: Tracy-release] Error 1
[19:49:15] make[1]: Leaving directory '/workspace/srcdir/tracy-0.7.8/profiler/build/unix'
```
Cool, I also got that at some point but I thought I messed up something (I wasn't using the 10.15 SDK, I thought that was the problem).  I have no clue what that means :slightly_smiling_face: 

Note that you have to manually install the license now.

**Comment by giordano** [comment]
Ok, it looks like we have a problem with link-time optimisation, but then after removing `-flto` I get
```
c++ -O3 -mmacosx-version-min=10.15 -std=c++17 -D"DISPLAY_SERVER_X11" -s -DNDEBUG -D__STDC_FORMAT_MACROS -DIMGUI_IMPL_OPENGL_LOADER_GL3W -DIMGUI_ENABLE_FREETYPE obj/release/o/o/o/../../../common/TracySocket.o obj/release/o/o/o/../../../common/TracySystem.o obj/release/o/o/o/../../../common/tracy_lz4.o obj/release/o/o/o/../../../common/tracy_lz4hc.o obj/release/o/o/o/../../../imgui/imgui.o obj/release/o/o/o/../../../imgui/imgui_demo.o obj/release/o/o/o/../../../imgui/imgui_draw.o obj/release/o/o/o/../../../imgui/imgui_tables.o obj/release/o/o/o/../../../imgui/imgui_widgets.o obj/release/o/o/o/../../../imgui/misc/freetype/imgui_freetype.o obj/release/o/o/o/../../../server/TracyBadVersion.o obj/release/o/o/o/../../../server/TracyColor.o obj/release/o/o/o/../../../server/TracyFilesystem.o obj/release/o/o/o/../../../server/TracyMemory.o obj/release/o/o/o/../../../server/TracyMicroArchitecture.o obj/release/o/o/o/../../../server/TracyMmap.o obj/release/o/o/o/../../../server/TracyMouse.o obj/release/o/o/o/../../../server/TracyPrint.o obj/release/o/o/o/../../../server/TracySourceContents.o obj/release/o/o/o/../../../server/TracySourceTokenizer.o obj/release/o/o/o/../../../server/TracySourceView.o obj/release/o/o/o/../../../server/TracyStackFrames.o obj/release/o/o/o/../../../server/TracyStorage.o obj/release/o/o/o/../../../server/TracyTaskDispatch.o obj/release/o/o/o/../../../server/TracyTexture.o obj/release/o/o/o/../../../server/TracyTextureCompression.o obj/release/o/o/o/../../../server/TracyThreadCompress.o obj/release/o/o/o/../../../server/TracyUserData.o obj/release/o/o/o/../../../server/TracyView.o obj/release/o/o/o/../../../server/TracyWorker.o obj/release/o/o/o/../../src/HttpRequest.o obj/release/o/o/o/../../src/imgui_impl_glfw.o obj/release/o/o/o/../../src/imgui_impl_opengl3.o obj/release/o/o/o/../../src/main.o obj/release/o/o/o/../../src/NativeWindow.o obj/release/o/o/o/../../src/ResolvService.o obj/release/o/o/o/../../src/winmain.o obj/release/o/o/o/../../src/winmainArchDiscovery.o obj/release/o/o/o/../../../nfd/nfd_common.o obj/release/o/o/o/../../../zstd/common/debug.o obj/release/o/o/o/../../../zstd/common/entropy_common.o obj/release/o/o/o/../../../zstd/common/error_private.o obj/release/o/o/o/../../../zstd/common/fse_decompress.o obj/release/o/o/o/../../../zstd/common/pool.o obj/release/o/o/o/../../../zstd/common/threading.o obj/release/o/o/o/../../../zstd/common/xxhash.o obj/release/o/o/o/../../../zstd/common/zstd_common.o obj/release/o/o/o/../../../zstd/compress/fse_compress.o obj/release/o/o/o/../../../zstd/compress/hist.o obj/release/o/o/o/../../../zstd/compress/huf_compress.o obj/release/o/o/o/../../../zstd/compress/zstdmt_compress.o obj/release/o/o/o/../../../zstd/compress/zstd_compress.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_literals.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_sequences.o obj/release/o/o/o/../../../zstd/compress/zstd_compress_superblock.o obj/release/o/o/o/../../../zstd/compress/zstd_double_fast.o obj/release/o/o/o/../../../zstd/compress/zstd_fast.o obj/release/o/o/o/../../../zstd/compress/zstd_lazy.o obj/release/o/o/o/../../../zstd/compress/zstd_ldm.o obj/release/o/o/o/../../../zstd/compress/zstd_opt.o obj/release/o/o/o/../../../zstd/decompress/huf_decompress.o obj/release/o/o/o/../../../zstd/decompress/zstd_ddict.o obj/release/o/o/o/../../../zstd/decompress/zstd_decompress.o obj/release/o/o/o/../../../zstd/decompress/zstd_decompress_block.o obj/release/o/o/o/../../../zstd/dictBuilder/cover.o obj/release/o/o/o/../../../zstd/dictBuilder/divsufsort.o obj/release/o/o/o/../../../zstd/dictBuilder/fastcover.o obj/release/o/o/o/../../../zstd/dictBuilder/zdict.o obj/release/o/o/o/../../libs/gl3w/GL/gl3w.o obj/release/o/o/o/../../../nfd/nfd_cocoa.o -L/workspace/destdir/lib -lglfw -lfreetype -lcapstone  -lpthread -ldl -framework CoreFoundation -framework AppKit -o Tracy-release
ld: warning: option -s is obsolete and being ignored
Undefined symbols for architecture x86_64:
  "__ZdaPvSt11align_val_t", referenced from:
      __ZN5tracy6WorkerC2ERNS_8FileReadENS_9EventType4TypeEb in TracyWorker.o
  "__ZdlPvSt11align_val_t", referenced from:
      __ZN5tracy4View11DrawCompareEv in TracyView.o
      __ZNSt3__120__shared_ptr_pointerIPN5tracy8FileReadENS_14default_deleteIS2_EENS_9allocatorIS2_EEE16__on_zero_sharedEv in TracyView.o
      _main in main.o
      __ZL12DrawContentsv in main.o
      __ZNSt3__120__shared_ptr_pointerIPN5tracy8FileReadENS_14default_deleteIS2_EENS_9allocatorIS2_EEE16__on_zero_sharedEv in main.o
  "__ZnamSt11align_val_t", referenced from:
      __ZN5tracy6WorkerC2ERNS_8FileReadENS_9EventType4TypeEb in TracyWorker.o
  "__ZnwmSt11align_val_t", referenced from:
      __ZN5tracy4View11DrawCompareEv in TracyView.o
      _main in main.o
      __ZL12DrawContentsv in main.o
  "_objc_alloc_init", referenced from:
      _NFD_OpenDialog in nfd_cocoa.o
      _AddFilterListToDialog in nfd_cocoa.o
      _NFD_OpenDialogMultiple in nfd_cocoa.o
      _NFD_SaveDialog in nfd_cocoa.o
      _NFD_PickFolder in nfd_cocoa.o
ld: symbol(s) not found for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [../../../common/unix.mk:61: Tracy-release] Error 1
```
Sigh

**Comment by giordano** [comment]
Am I correct to say that GTK isn't used in macOS? https://github.com/wolfpld/tracy/blob/199b6ba7987d0c66cc298dd192ea270b24828d76/profiler/build/unix/build.mk#L25-L33

**Comment by jakebolewski** [comment]
Yes that is right, on linux it users [NativeFileDialog](https://github.com/mlabbe/nativefiledialog) but on OSX it looks to just uses the system file dialog.

I didn't know how to have a target only build dependency, still needed to read up on that.

File: T/TracyProfiler/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    ExecutableProduct("capture-release", :tracy_capture_exe),
+    ExecutableProduct("csvexport-release", :tracy_csvexport_exe),
+    ExecutableProduct("import-chrome-release", :tracy_import_chrome_exe),
+]
+
+dependencies = [
+    Dependency("GLFW_jll"),
+    Dependency("FreeType2_jll"),
+    Dependency("Capstone_jll"),
+    Dependency("GTK3_jll"),
```
Comment (giordano) [inline T/TracyProfiler/build_tarballs.jl:89]: This is how to set platform-dependent platforms
```suggestion
    Dependency("GTK3_jll"; platforms=filter(Sys.islinux, platforms)),
```
I don't know if there are more here

Comment (jakebolewski) [reply T/TracyProfiler/build_tarballs.jl:89]: I'll give it a try, thanks!

**Comment by jakebolewski** [comment]
Thanks for all your help and patience as I fumble trying to learn how to do this @giordano

**Comment by giordano** [comment]
> I didn't know how to have a target only build / runtime dependency, still needed to read up on that.

Your comment reminded me that we didn't really have much information about platform-specific dependencies, besides docstrings of dependency constructors.  I added a new section to the documentation: https://docs.binarybuilder.org/latest/building/#Platform-dependent-dependencies

**Pull request merged**